logstash-core/src/main/java/org/logstash/instrument/metrics/timer/ExecutionMillisTimer.java

```diff
@@ -0,0 +1,244 @@
+    @Override
+    public ExecutionCommitter begin() {
+        final long newCheckpointNanos = trackedMillisAdjustment.updateAndGet(MillisAdjustmentState::withIncrementedConcurrency).getCheckpointNanoTime();
+        return new ExecutionCommitter(newCheckpointNanos, () -> trackedMillisAdjustment.updateAndGet(MillisAdjustmentState::withDecrementedConcurrency).getCheckpointNanoTime());
+    }
+
+    @Override
+    public void reportUntracked(final long millis) {
+        untrackedMillis.add(millis);
+    }
+
+    @Override
+    public MetricType getType() {
+        return MetricType.COUNTER_LONG;
+    }
+
+    @Override
+    public Long getValue() {
+        return trackedMillisAdjustment.get().adjust(untrackedMillis.longValue());
+    }
+
+    public static ExecutionMillisTimer fromRubyBase(final AbstractNamespacedMetricExt metric,
+                                                    final RubySymbol key) {
+        final ThreadContext context = RubyUtil.RUBY.getCurrentContext();
+        final IRubyObject timer = metric.timer(context, key);
+
+        if (ExecutionMillisTimer.class.isAssignableFrom(timer.getJavaClass())) {
+            return timer.toJava(ExecutionMillisTimer.class);
+        } else {
+            return NULL_TIMER;
+        }
+    }
+
+    static class ExecutionCommitter implements Committer {
+        private final long startNanos;
+        private final LongSupplier atomicCommitAction;
+        private final AtomicBoolean committed = new AtomicBoolean(false);
+
+        /**
+         *
+         * @param startNanos a marker for the start time
+         * @param atomicCommitAction an action that will be executed exactly once,
+         *                           which returns the number of NANOSECONDS since the provided start time.
+         */
+        ExecutionCommitter(final long startNanos, final LongSupplier atomicCommitAction) {
+            this.startNanos = startNanos;
+            this.atomicCommitAction = atomicCommitAction;
+        }
+        @Override
+        public long commit() {
+            if (committed.compareAndSet(false, true)) {
+                final long commitNanos = atomicCommitAction.getAsLong();
+                return TimeUnit.NANOSECONDS.toMillis(commitNanos - this.startNanos);
+            } else {
+                return 0L;
+            }
+        }
+    }
+
+    interface MillisAdjustmentState {
+        long getCheckpointNanoTime();
+        MillisAdjustmentState withIncrementedConcurrency();
+        MillisAdjustmentState withDecrementedConcurrency();
+        long adjust(long externalMillis);
+    }
+
+    /**
+     * A {@link StaticAdjustmentState} is static, in that it is not tracking
+     * any concurrent executions, which makes calculations static and enables
+     * us to guard against negative concurrency.
+     */
+    private class StaticAdjustmentState implements MillisAdjustmentState {
+        private final long checkpointNanos;
+        private final long cumulativeMillis;
```

Comment:
clarity: this is the number of cumulative millis that had elapsed as of our checkpoint's _timestamp_.
```suggestion
        private final long committedMillis;
```